modules/tinymce/src/core/test/ts/browser/util/ImageUploaderTest.ts

```diff
@@ -0,0 +1,59 @@
+import { Log, Pipeline, Chain } from '@ephox/agar';
+import { Assert, UnitTest } from '@ephox/bedrock-client';
+import { ApiChains, TinyLoader } from '@ephox/mcagar';
+import { Arr } from '@ephox/katamari';
+import ImageUploader, { UploadResult } from 'tinymce/core/api/util/ImageUploader';
+
+import { BlobCache, BlobInfo } from 'tinymce/core/api/file/BlobCache';
+
+import Theme from 'tinymce/themes/silver/Theme';
+
+UnitTest.asynctest('browser.tinymce.core.util.ImageUploaderTest', (success, failure) => {
+
+  Theme();
+
+  TinyLoader.setupLight((editor, onSuccess, onFailure) => {
+    const cache = BlobCache();
+    const blob = new Blob([ JSON.stringify({ hello: 'world' }) ]);
+
+    const image1 = cache.create({ blob, base64: 'test' });
+
+    const cUploadImages = (images: BlobInfo[], notifications: boolean) => Chain.async((_value, next, die) => {
+      const uploader = ImageUploader(editor);
+      uploader.upload(images, notifications).then(next, die);
+    });
+
+    const cAssertUploadResultSuccess = (expectedLength: number, url: string) => Chain.op((uploadResults: UploadResult[]) => {
+      Assert.eq('Number of results', expectedLength, uploadResults.length);
+      Arr.each(uploadResults, (uploadResult) => {
+        Assert.eq('Url is Image.png', url, uploadResult.url);
+        Assert.eq('Successful', true, uploadResult.status);
+      });
+    });
+
+    const cAssertUploadResultFailure = (expectedLength: number, url: string) => Chain.op((uploadResults: UploadResult[]) => {
+      Assert.eq('Number of results', expectedLength, 0);
```

Comment:
This just asserts that the 0 is 0?

Reply:
The expected length of results is 0

Reply:
Right. But if the `expectedLength` is 0, would you not want to compare that to the actual length, not 0?

```suggestion
      Assert.eq('Number of results', expectedLength, uploadResults.length);
```

Reply:
I think that maybe because we are constructing a BlobCache the array will not be empty actually, it will return this actually.

![Screen Shot 2020-11-13 at 8 59 49 am](https://user-images.githubusercontent.com/6000860/99006842-dee40000-258e-11eb-84dd-41c6a9bb2094.png)

Reply:
I think that this is actually the proper way...

![Screen Shot 2020-11-13 at 9 26 09 am](https://user-images.githubusercontent.com/6000860/99008661-52d3d780-2592-11eb-9c72-bc6d8490ed2b.png)

Reply:
I think the API of the two assert functions should match, so the first parameters is the expected length of the result and the second is the expected URL.

In your screenshot here you're comparing the length of the URL to the expected length, but then you also compare the expected `url` to an empty string, which will always be true.

It might be better to rename your `url` variable to `expectedUrl` to avoid confusion?

Reply:
I also think it's worth asserting the `error` as that's what is being set in `uploadImage` and in your test when you call `failure('Error')`

Reply:
Good ones.... let me refactor it.

Reply:
@JamesToohey I just pushed some changes, please have a look.